### PR TITLE
Add support for structs in interface block and Multidimensional Arrays 

### PR DESCRIPTION
**Pull request by spencer-lunarg**
There is at least more commit to follow this (to better organize the code and remove the `bool nested_struct` workaround), but this adds many more tests. It also fixes issues having structs in the interface blocks as well as having  Multidimensional Arrays interface

### CODE REVIEW

File: layers/state_tracker/shader_module.cpp [hunk 1]
```diff
@@ -1211,6 +1211,7 @@ uint32_t SHADER_MODULE_STATE::GetComponentsConsumedByType(uint32_t type) const {
             return GetComponentsConsumedByType(insn->Word(3));
         case spv::OpTypeArray:
             // Skip array as each array element is a whole new Location and we care only about the base type
+            // ex. vec3[5] will only return 3
```
Comment (artem-lunarg) [inline layers/state_tracker/shader_module.cpp:1214]: nice example

File: layers/state_tracker/shader_module.cpp [hunk 2]
```diff
@@ -1478,19 +1479,24 @@ bool StageInteraceVariable::IsArrayInterface(const StageInteraceVariable& variab
...
     // This is allowed only here because interface variables are never Phyiscal pointers
-    const Instruction& base_type = *module_state.FindDef(module_state.FindDef(variable.type_id)->Word(3));
+    const Instruction* base_type = module_state.FindDef(module_state.FindDef(variable.type_id)->Word(3));
 
-    // Strip away the array if one
-    if (variable.is_array_interface && (base_type.Opcode() == spv::OpTypeArray || base_type.Opcode() == spv::OpTypeRuntimeArray)) {
-        const uint32_t type_id = base_type.Word(2);
-        return *module_state.FindDef(type_id);
-    }
+    // Strip away the first array, if any, if speical interface array
```
Comment (artem-lunarg) [inline layers/state_tracker/shader_module.cpp:1487]: typo: speical

File: layers/state_tracker/shader_module.cpp [hunk 3]
```diff
@@ -1537,46 +1581,51 @@ std::vector<InterfaceSlot> StageInteraceVariable::GetInterfaceSlots(StageInterac
...
             for (uint32_t i = 0; i < variable.type_struct_info->length; i++) {
-                const uint32_t member_id = variable.type_struct_info->members[i].id;
+                const auto& member = variable.type_struct_info->members[i];
+                const uint32_t member_id = member.id;
                 // Location/Components cant be decorated in nested structs, so no need to keep checking further
                 // The spec says all or non of the member variables must have Location
                 const auto member_decoration = variable.type_struct_info->decorations.member_decorations.at(i);
-                const uint32_t location = member_decoration.location;
+                uint32_t location = member_decoration.location;
                 const uint32_t starting_componet = member_decoration.component;
```
Comment (artem-lunarg) [inline layers/state_tracker/shader_module.cpp:1590]: if you are going to update previous typo please update this one: starting_componet

**Review by artem-lunarg — APPROVED** [review]

File: layers/state_tracker/shader_module.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply layers/state_tracker/shader_module.cpp:1590]: .... I rather just fix it now, CI is not busy right luckily... curse those :gb: and their English :sob:

Comment (spencer-lunarg) [reply layers/state_tracker/shader_module.cpp:1590]: also makes `ctrl+f` much easier for the future

**Pull request merged**